File: leetCode/fukun/201-300/lc_208.js
function MultiTreeNode(val) {
    this.val = val;
    this.next = [];
}

var Trie = function () {
    this.root = new MultiTreeNode(null);
};

Trie.prototype.insert = function (word) {
    word += '0';
    let curr = 0,
        currT = this.root;
    sync: while (1) {
        for (let i = 0; i < currT.next.length; i++) {
            if (currT.next[i].val === word[curr]) {
                currT = currT.next[i];
                curr++;
                continue sync;
            }
        }
        for (let i = curr; i < word.length; i++) {
            let node = new MultiTreeNode(word[i]);
            currT.next.push(node);
            currT = node;
        }
        break sync;
    }
};

Trie.prototype.search = function (word) {
    word += '0';
    let curr = 0,
        currT = this.root;
    sync:while (1) {
        if (currT.next.length === 0 && curr === word.length) return true;
        for (let i = 0; i < currT.next.length; i++) {
            if (currT.next[i].val === word[curr]) {
                currT = currT.next[i];
                curr++;
                continue sync;
            }
        }
        return false;
    }
};

Trie.prototype.startsWith = function (prefix) {
    let curr = 0,
        currT = this.root;
    sync:while (1) {
        if (curr === prefix.length) return true;
        for (let i = 0; i < currT.next.length; i++) {
            if (currT.next[i].val === prefix[curr]) {
                currT = currT.next[i];
                curr++;
                continue sync;
            }
        }
        return false;
    }
};

var obj = new Trie();
obj.insert('leet');
obj.insert('code');
console.log(obj.search('le'))
console.log(obj.startsWith('le'))
console.log(obj.search('coet'))

File: leetCode/lwc/0201-0300/lc_208.java
package leetcode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

/**
 * @author lwc
 * @date 2019/11/19 16:09
 */
public class lc_208 {
    class Trie {
        class TrieNode {
            TrieNode[] nodes;
            TrieNode() {
                nodes = new TrieNode[26];
            }
        }

        TrieNode root;
        HashSet<String> words;
        /** Initialize your data structure here. */
        public Trie() {
            root = new TrieNode();
            words = new HashSet<>();
        }

        /** Inserts a word into the trie. */
        public void insert(String word) {
            if(!words.contains(word)) {
                words.add(word);
                char[] arr = word.toCharArray();
                TrieNode curr = root;
                for (char ch : arr) {
                    TrieNode node = curr.nodes[ch - 'a'];
                    if (node == null) {
                        curr.nodes[ch - 'a'] = new TrieNode();
                    }
                    curr = curr.nodes[ch - 'a'];
                }
            }
        }

        /** Returns if the word is in the trie. */
        public boolean search(String word) {
            return words.contains(word);
        }

        /** Returns if there is any word in the trie that starts with the given prefix. */
        public boolean startsWith(String prefix) {
            char[] arr = prefix.toCharArray();
            TrieNode curr = root;
            for (char ch: arr) {
                TrieNode node = curr.nodes[ch - 'a'];
                if (node == null) {
                    return false;
                }
                curr = node;
            }
            return true;
        }
    }
}
